tests/Java.Interop-Tests/Java.Interop/JavaPeerableExtensionsTests.cs

```diff
@@ -0,0 +1,114 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using Java.Interop;
+
+using NUnit.Framework;
+
+namespace Java.InteropTests;
+
+[TestFixture]
+public class JavaPeerableExtensionsTests {
+
+	[Test]
+	public void JavaAs_Exceptions ()
+	{
+		using var v = new MyJavaInterfaceImpl ();
+
+		// The Java type corresponding to JavaObjectWithMissingJavaPeer doesn't exist
+		Assert.Throws<ArgumentException>(() => v.JavaAs<JavaObjectWithMissingJavaPeer>());
+
+		var r = v.PeerReference;
+		using var o = new JavaObject (ref r, JniObjectReferenceOptions.Copy);
+		// MyJavaInterfaceImpl doesn't provide an activation constructor
+		Assert.Throws<NotSupportedException>(() => o.JavaAs<MyJavaInterfaceImpl>());
+#if !__ANDROID__
+		// JavaObjectWithNoJavaPeer has no Java peer
+		Assert.Throws<ArgumentException>(() => v.JavaAs<JavaObjectWithNoJavaPeer>());
+#endif  // !__ANDROID__
+	}
+
+	[Test]
+	public void JavaAs_NullSelfReturnsNull ()
+	{
+		Assert.AreEqual (null, JavaPeerableExtensions.JavaAs<IAndroidInterface> (null));
+	}
+
+	public void JavaAs_InvalidPeerRefReturnsNull ()
+	{
+		var v   = new MyJavaInterfaceImpl ();
+		v.Dispose ();
+		Assert.AreEqual (null, JavaPeerableExtensions.JavaAs<IJavaInterface> (v));
+	}
+
+	[Test]
+	public void JavaAs_InstanceThatDoesNotImplementInterfaceReturnsNull ()
+	{
+		using var v = new MyJavaInterfaceImpl ();
+		Assert.AreEqual (null, JavaPeerableExtensions.JavaAs<IAndroidInterface> (v));
+	}
```

Comment:
Do we need a test that tries to cast to a random, incorrect type like `System.Guid`? (One that isn't even a Java thing) Would that case throw?

Reply:
I believe the case you are suggesting is compiler prevented by the generic constraint?

```
public static TResult? JavaAs<TResult> (this IJavaPeerable? self)
    where TResult : class, IJavaPeerable
{ ... }
```

Or perhaps I misunderstand your comment.